### PR TITLE
PYIC-4116: Added Address CRI consumer contract tests

### PR DESCRIPTION
**Pull request by RogVanGDS**
### What changed

Added Address CRI consumer contract tests

### Why did it change

To build up contract test coverage for Address CRI service, and unblock producer side tests for Address CRI

### Issue tracking

- [PYIC-4116](https://govukverify.atlassian.net/browse/PYIC-4116)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

[PYIC-4116]: https://govukverify.atlassian.net/browse/PYIC-4116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+                                throw new RuntimeException(e);
+                            }
+                        });
+    }
+
+    @Pact(provider = "AddressCriProvider", consumer = "IpvCoreBack")
+    public RequestResponsePact invalidCredentialsRequestThrowsError(PactDslWithProvider builder)
+            throws Exception {
+        return builder.given("dummyApiKey is a valid api key")
+                .given("dummyInvalidAccessToken is a valid access token")
```
Comment (DanCorderIPV) [inline lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:561]: `is a valid` -> `is an invalid`

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+        when(mockConfigService.getCriPrivateApiKey(any())).thenReturn(PRIVATE_API_KEY);
+
+        // We need to generate a fixed request, so we set the secure token and expiry to constant
+        // values.
+        var underTest =
+                new CriApiService(
+                        mockConfigService, mockSigner, mockSecureTokenHelper, CURRENT_TIME);
+
+        // Act
+
```
Comment (DanCorderIPV) [inline lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java]: blank line

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+    // sent by the CRI team
+    private static final String VALID_VC_HEADER =
+            """
+            {
+              "typ": "JWT",
+              "alg": "ES256"
+            }
+            """;
+    // 2099-01-01 00:00:00 is 4070908800 in epoch seconds
+    private static final String VALID_EXPERIAN_ADDRESS_VC_BODY =
```
Comment (DanCorderIPV) [inline lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:93]: I see that these have different address fields, but why is one `Experian` and the other not?

**Review by DanCorderIPV — DISMISSED** [review]

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java (hunk 3 above)
Comment (RogVanGDS) [reply lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:93]: I believe sometimes the address is mapped to match the fields Experian Fraud check is expecting for fraud checking reasons, not 100% sure on this though

**Review by DanCorderIPV — APPROVED** [review]

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java (hunk 3 above)
Comment (blakeyp) [reply lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:93]: The address CRI is really for just collecting the user's 'claimed' address which is then used in the fraud check

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+    }
+
+    @Pact(provider = "AddressCriProvider", consumer = "IpvCoreBack")
+    public RequestResponsePact invalidAuthCodeReturnsTokenError(PactDslWithProvider builder) {
+        return builder.given("dummyInvalidAuthCode is an invalid authorization code")
+                .given("dummyApiKey is an invalid api key")
+                .given("dummyAddressComponentId is the address CRI component ID")
+                .given(
+                        "Address CRI uses CORE_BACK_SIGNING_PRIVATE_KEY_JWK to validate core signatures")
+                .uponReceiving("Valid auth code")
```
Comment (blakeyp) [inline lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:281]: Shouldn't this be 'valid api key' but 'invalid auth code'?

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+                .method("POST")
+                .body(
+                        "client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&code=dummyInvalidAuthCode&grant_type=authorization_code&redirect_uri=https%3A%2F%2Fidentity.staging.account.gov.uk%2Fcredential-issuer%2Fcallback%3Fid%3Daddress&client_assertion=eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJpc3MiOiJpcHYtY29yZSIsInN1YiI6Imlwdi1jb3JlIiwiYXVkIjoiZHVtbXlBZGRyZXNzQ29tcG9uZW50SWQiLCJleHAiOjQwNzA5MDk3MDAsImp0aSI6IlNjbkY0ZEdYdGhaWVhTXzVrODVPYkVvU1UwNFctSDNxYV9wNm5wdjJaVVkifQ.hXYrKJ_W9YItUbZxu3T63gQgScVoSMqHZ43UPfdB8im8L4d0mZPLC6BlwMJSsfjiAyU1y3c37vm-rV8kZo2uyw")
+                .headers(
+                        "x-api-key",
+                        PRIVATE_API_KEY,
+                        "Content-Type",
+                        "application/x-www-form-urlencoded; charset=UTF-8")
+                .willRespondWith()
+                .status(401)
```
Comment (blakeyp) [inline lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:292]: Is it 401 or 400 - for other CRIs it seems to be 400?

File: lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java (hunk 4 above)
Comment (RogVanGDS) [reply lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:281]: This has already been changed in the code (The PR is out of date)

Comment (blakeyp) [reply lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:281]: I think it still says "dummyApiKey is an invalid api key" ?

Comment (RogVanGDS) [reply lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java:281]: Oh right I see what you mean now

**Pull request merged**